### PR TITLE
Updating TF pets training example

### PR DESCRIPTION
**Pull request by SundarRajan28**
- Updated tensorflow dockerfile and removed cupy and tf-models package installation
- Updated TF pets training example to use latest TF2 training APIs
- Added tfrecord creation script and updated README with training instructions

### CODE REVIEW

File: docs/examples/tf/pets_training/create_pet_tf_record.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (LakshmiKumar23) [inline docs/examples/tf/pets_training/create_pet_tf_record.py:3]: if this is Apache license, we cannot host this file. We should give a link to the tf repo and ask user to run from there. 
@kiritigowda can you confirm this?

File: docs/examples/tf/pets_training/pet_label_map.pbtxt [hunk 2]
```diff
@@ -0,0 +1,184 @@
+item {
```
Comment (LakshmiKumar23) [inline docs/examples/tf/pets_training/pet_label_map.pbtxt:1]: is this file created using the ```download_and_preprocess_dataset.sh``` script. If yes, why are we checking this file in?

File: docs/examples/tf/pets_training/requirements.sh [hunk 3]
```diff
@@ -1,9 +1,2 @@
 # installing required packages
-pip install contextlib2
-git clone https://github.com/tensorflow/models.git
-cd models/research
-protoc object_detection/protos/*.proto --python_out=.
-cp object_detection/packages/tf2/setup.py .
-python -m pip install .
-pip install tensorflow==2.11.0
-cd ../../
+pip install contextlib2 pillow
```
Comment (rrawther) [inline docs/examples/tf/pets_training/requirements.sh:2]: please add this single line in the download_preprocess_...sh and get rid of this file

File: docker/rocal-with-tensorflow.dockerfile [hunk 4]
```diff
@@ -56,14 +52,3 @@ WORKDIR $ROCAL_WORKSPACE
 RUN pip install --upgrade pip
 RUN git clone -b develop https://github.com/ROCm/rocAL && \
         mkdir build && cd build && cmake -D PYTHON_VERSION_SUGGESTED=${ROCAL_PYTHON_VERSION_SUGGESTED} ../rocAL && make -j8 && cmake --build . --target PyPackageInstall && make install
-
-WORKDIR /workspace
-
-#install tensorflow dependencies - Compile protos and Install TensorFlow Object Detection API.
-RUN cd ~/models/research/ && \
```
Comment (LakshmiKumar23) [inline docker/rocal-with-tensorflow.dockerfile:63]: if you need this for ```create_pet_tf_record.py```, our TF dockers already comes with ```models/research```. Only location change from ~/. Please recheck

**Review by rrawther — CHANGES_REQUESTED** [review]
Added comments.

File: docs/examples/tf/pets_training/pet_label_map.pbtxt (hunk 2 above)
Comment (SundarRajan28) [reply docs/examples/tf/pets_training/pet_label_map.pbtxt:1]: This file is used in create_pet_tf_record.py to create the TFrecords. It is needed

File: docker/rocal-with-tensorflow.dockerfile (hunk 4 above)
Comment (SundarRajan28) [reply docker/rocal-with-tensorflow.dockerfile:63]: The code in create_pet_tf_record.py is adapted from this [file] (https://github.com/tensorflow/models/blob/master/research/object_detection/dataset_tools/create_pet_tf_record.py) so we no longer need this models package installation which causes GPU issues

File: docs/examples/tf/pets_training/create_pet_tf_record.py (hunk 1 above)
Comment (SundarRajan28) [reply docs/examples/tf/pets_training/create_pet_tf_record.py:3]: The code in this file is copied and modified from this [file](https://github.com/tensorflow/models/blob/master/research/object_detection/dataset_tools/create_pet_tf_record.py) so we no longer need this models package installation which causes GPU issues. I can remove this copyright if needed

File: docs/examples/tf/pets_training/download_and_preprocess_dataset.sh [hunk 5]
```diff
@@ -4,6 +4,7 @@ DATASET_URL="http://www.robots.ox.ac.uk/~vgg/data/pets/data/images.tar.gz"
 GROUNDTRUTH_URL="http://www.robots.ox.ac.uk/~vgg/data/pets/data/annotations.tar.gz"
 
 apt-get install wget
+pip install contextlib2 pillow
```
Comment (LakshmiKumar23) [inline docs/examples/tf/pets_training/download_and_preprocess_dataset.sh:7]: please add ```lxml``` to this

**Comment by LakshmiKumar23** [comment]
@SundarRajan28 Fails to run. I have tf for rocm built on bare-metal. 
```
lakshmi@kapu:~/work/rocAL/docs/examples/tf/pets_training$ python3 train.py

-----------------------------------------------------------------------------------------
TF records (train) are located in tf_pets_records/train/
TF records (val) are located in tf_pets_records/val/
-----------------------------------------------------------------------------------------

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #
=================================================================
 mobilenetv2_1.00_128 (Func  (None, 4, 4, 1280)        2257984
 tional)

 global_average_pooling2d (  (None, 1280)              0
 GlobalAveragePooling2D)

 dense (Dense)               (None, 37)                47397

=================================================================
Total params: 2305381 (8.79 MB)
Trainable params: 47397 (185.14 KB)
Non-trainable params: 2257984 (8.61 MB)
_________________________________________________________________
OK: loaded 125 kernels from libvx_rpp.so
Pipeline has been created succesfully
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00006-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00009-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00004-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00008-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00007-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00003-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00000-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00005-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00002-of-00010
Reading for image classification - file_name:: tf_pets_records/train//pet_faces_train.record-00001-of-00010
OK: loaded 125 kernels from libvx_rpp.so
Pipeline has been created succesfully
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00007-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00008-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00000-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00009-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00003-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00001-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00006-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00005-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00004-of-00010
Reading for image classification - file_name:: tf_pets_records/val//pet_faces_val.record-00002-of-00010
info: Import CuPy failed. Falling back to CPU!
info: Import CuPy failed. Falling back to CPU!
Epoch : 1
error: Failure when generating HSACO
Traceback (most recent call last):
  File "/home/lakshmi/work/rocAL/docs/examples/tf/pets_training/train.py", line 182, in <module>
    main()
  File "/home/lakshmi/work/rocAL/docs/examples/tf/pets_training/train.py", line 157, in main
    gradients = tape.gradient(loss, model.trainable_variables)
  File "/home/lakshmi/.local/lib/python3.10/site-packages/tensorflow/python/eager/backprop.py", line 1066, in gradient
    flat_grad = imperative_grad.imperative_grad(
  File "/home/lakshmi/.local/lib/python3.10/site-packages/tensorflow/python/eager/imperative_grad.py", line 67, in imperative_grad
    return pywrap_tfe.TFE_Py_TapeGradient(
  File "/home/lakshmi/.local/lib/python3.10/site-packages/tensorflow/python/eager/backprop.py", line 148, in _gradient_function
    return grad_fn(mock_op, *out_grads)
  File "/home/lakshmi/.local/lib/python3.10/site-packages/tensorflow/python/ops/math_grad.py", line 1507, in _DivNoNanGrad
    gy = grad * math_ops.div_no_nan(math_ops.div_no_nan(-x, y), y)
  File "/home/lakshmi/.local/lib/python3.10/site-packages/tensorflow/python/ops/gen_math_ops.py", line 7007, in neg
    _ops.raise_from_not_ok_status(e, name)
  File "/home/lakshmi/.local/lib/python3.10/site-packages/tensorflow/python/framework/ops.py", line 5883, in raise_from_not_ok_status
    raise core._status_to_exception(e) from None  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.UnknownError: {{function_node __wrapped__Neg_device_/job:localhost/replica:0/task:0/device:GPU:0}} JIT compilation failed. [Op:Neg] name:

```

**Review by rrawther — APPROVED** [review]
Can be merged after all comments are resolved

**Comment by SundarRajan28** [comment]
@LakshmiKumar23 The import cupy failure messages are due to this [line](https://github.com/ROCm/rocAL/blob/develop/rocAL_pybind/amd/rocal/plugin/tf.py#L99). This condition will always be True. The correct condition should be `if device in ("gpu", "cuda")`. This would only check the cupy import when its a GPU device.
The errors printed in the above log when running the TF pets training must be due to a rocm-tensorflow issue as evident by this line
`error: Failure when generating HSACO`
The backtrace is also pointing to the tape.gradients line which means this is a TF related error. Can you try any other TF example training in the bare metal to check if TF works on bare metal?

**Review by LakshmiKumar23 — APPROVED** [review]
Works on docker. Issues only with bare-metal. Can be merged

**Review by rrawther — APPROVED** [review]

**Pull request merged**